plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -212,17 +213,123 @@ processTerm = handleTerm <=< traverseOf termSubtypes applyTypeSubstitution where
             -- actually have got rid of all of them!
             pure $ mkLet ann NonRec bs' t'
         -- This includes recursive let terms, we don't even consider inlining them at the moment
-        t ->
-            -- process all subterms first, so that the rhs won't be processed more than once. This
-            -- is important because otherwise the number of times we process them can grow
-            -- exponentially in the case that it has nested `let`s.
-            --
-            -- Then, consider call site inlining for each node that have gone through unconditional
-            -- inlining. Because `inlineSaturatedApp` traverses *all* application nodes for each
-            -- subterm, the runtime is quadratic for terms with a long chain of applications.
-            -- If we use the context-based approach like in GHC, this won't be a problem, so we may
-            -- consider that in the future.
-            inlineSaturatedApp =<< forMOf termSubterms t processTerm
+        t -> do
+            -- See note [Processing order of call site inlining]
+            let (hd, args) = splitApplication t
+                -- process the term args, ignore the type arguments
+                processArgs :: AppContext tyname name uni fun ann
+                    -> InlineM tyname name uni fun ann (AppContext tyname name uni fun ann)
+                processArgs (TermAppContext arg ann ctx) = do
+                    processedArg <- processTerm arg
+                    processedArgs <- processArgs ctx
+                    pure $ TermAppContext processedArg ann processedArgs
+                processArgs (TypeAppContext ty ann ctx) = do
+                    processedArgs <- processArgs ctx
+                    pure $ TypeAppContext ty ann processedArgs
+                processArgs AppContextEnd = pure AppContextEnd
+            case (hd, args) of
+                (Var{}, _) -> do
+                    processedHd <- processTerm hd -- the variable may be unconditionally inlined.
+                    case processedHd of
+                        -- if it didn't get unconditionally inlined/got inlined to another variable
+                        (Var _ name) -> do
+                                gets (lookupVarInfo name) >>= \case
+                                    Just varInfo -> do
+                                        -- process the args
+                                        processedArgs <- processArgs args
+                                        maybeInlined <- callSiteInline t varInfo processedArgs
```

Comment:
In fact you don't need to pass a term. You can just pass the size, i.e., `termSize (fillAppContext processedHd processedArgs)`.

Reply:
In https://github.com/input-output-hk/plutus/pull/5426/commits/e71d7dd4c0a00f75f45fc2ce5f6b6df5c182f7cd I check the term size of `forMOf termSubterms t processTerm` because as I explained when we're not inlining we're returning `forMOf termSubterms t processTerm`. This actually make it so that the script context tests are not impacted by this change. So it's very curious how when comparing the sizes the other way actually caused some improvements for the other script context tests - this actually seems a rather strong indicator that the current size metric is not very good. (The other tests that got improvements are measuring the budget, while the `fitsInto` tests were using the same size metric).